File: 4/4.25-2.swift

// どうしても可変な設計にする場合

import Foundation

class HitPoint {
    private static let min = 0

    // 可変であることが前提なのでletじゃないことはBADではない
    // [GOOD]
    // setterだけprivateにして外部からgetだけできるようにする
    private(set) var amount: Int

    init(amount: Int) throws {
        guard amount >= HitPoint.min else {
            throw fatalError()
        }

        self.amount = amount
    }

    func damege(damegeAmount: Int) {
        let nextAmount = amount - damegeAmount
        amount = max(HitPoint.min, nextAmount)
    }

    var isMin: Bool {
        amount == HitPoint.min
    }
}

enum State {
    case alive, death, poison
}

class Member {
    let hitPoint: HitPoint
    // Swiftの場合はletだと配列の中身が修正されることも許していないため
    // あとから状態を変えるのであればvarとなる
    // [GOOD]
    // setterだけprivateにして外部からgetだけできるようにする
    private(set) var states: [State]

    init(hitPoint: HitPoint, states: [State]) {
        self.hitPoint = hitPoint
        self.states = states
    }

    func damege(damegeAmount: Int) {
        hitPoint.damege(damegeAmount: damegeAmount)
        if hitPoint.isMin {
            states.removeAll { state in
                state == .alive
            }
            states.append(.death)
        }
    }
}

let member = Member(hitPoint: try! HitPoint(amount: 10), states: [State.alive])

member.damege(damegeAmount: 20)

// [GOOD]
// hitPointがマイナスにならない
// statesが切り替わっている
print(member.hitPoint.amount)
print(member.states)

// [GOOD]
// member.amountが外部から修正できない(ビルドが通らない)
// member.stateが外部から修正できない(ビルドが通らない)
member.hitPoint.amount = -20
member.states.append(.alive)
print(member.hitPoint.amount)
print(member.states)

File: 4/4.25-2.dart
import 'dart:math';

// どうしても可変な設計にする場合

class HitPoint {
  static final _MIN = 0;

  // 可変であることが前提なのでfinalじゃないことはBADではない
  // [GOOD]
  // privateなので外部から変更できない
  late int _amount;
  //getterだけ用意してamountにアクセスだけできるようにする
  int get amount {
    return _amount;
  }

  HitPoint(int amount) {
    if (amount < _MIN) {
      throw Exception();
    }
    this._amount = amount;
  }

  void damege(final int damegeAmount) {
    final nextAmount = _amount - damegeAmount;
    _amount = max(_MIN, nextAmount);
  }

  bool get isMin {
    return this._amount == _MIN;
  }
}

enum State { alive, death, poison }

class Member {
  final HitPoint hitPoint;
  // [注意]
  // Dartの場合finalだとして配列は中身を変えることができてしまう
  // [BAD]
  // 外部から変更できてしまう
  // Dartの配列は参照を渡すだけなのでHitPoint.amountのような対応もできない
  // 言語仕様上、注意がいる
  final List<State> states; //List.unmodifiable(states);

  Member(HitPoint this.hitPoint, List<State> this.states);

  void damege(int damegeAmount) {
    hitPoint.damege(damegeAmount);
    if (hitPoint.isMin) {
      
      states.remove(State.alive);
      states.add(State.death);
      // [おまけ]
      // こうやって..とするとメソッドチェーンみたいに書ける
//       states.remove(State.alive)
//         ..add(State.death);
    }
  }
}

void main() {
  final Member member = Member(HitPoint(10), [State.alive]);

  member.damege(20);

  // [GOOD]
  // hitPointがマイナスにならない
  // statesが切り替わっている
  print(member.hitPoint.amount);
  print(member.states);

  // [GOOD]
  // member.amountが外部から修正できない(ビルドが通らない)
  member.hitPoint.amount = -20;
  print(member.hitPoint.amount);
  // [BAD]
  // member.stateが外部に公開されているかつListのため、外部からも変更できてしまう
  member.states.add(State.alive);
  print(member.states);
}
